Adding regex predicate support

kishoreg opened the pull request:
Adding support for the REGEXP_LIKE operator. 

REGEXP_LIKE operator supports java style regular expressions. All comparisons are case insensitive by default. One can turn on/off flags by using [Embedded Flag Expressions](https://docs.oracle.com/javase/tutorial/essential/regex/pattern.html) 

e.g queries
```
select l_comment from tpch_lineitem where regexp_like (l_comment, '.*fox.*')
```

Note the initial commit supported MySQL  LIKE semantics but it was getting complicated to convert a LIKE expression to java regex expression and hence dropped it in the subsequent commit. Note the pattern does not have to match the entire string. If users want to match the entire string, they can use ^ at the start and $ at the end of the expression.

jfim commented on pinot-common/src/main/antlr4/com/linkedin/pinot/pql/parsers/PQL2.g4:
```diff
@@ -87,6 +88,9 @@ comparisonOperator: '<' | '>' | '<>' | '<=' | '>=' | '=';
   expression BETWEEN expression AND expression;
 
+regexClause:
+  expression (LIKE | REGEXP ) '(' literal (',' literal)* ')';
```
It would be good to not require parentheses for this. For example, in pretty much every SQL database, the syntax `foo LIKE 'a%'` is valid.

This also copies the (non-standard) `REGEXP` keyword from MySQL. Unfortunately, that extension doesn't support adding regexp flags (eg. case insensitive matching, multiline match, etc.). I took a quick look at how other databases do it:

- MySQL: `foo REGEXP 'a.*'`, no support for regular expression flags
- PostgreSQL: `foo ~ 'a.*'` for case sensitive match, `foo ~* 'a.*'` for case insentitive match, 
- MSSQL: Not supported
- Oracle: `REGEXP_LIKE(foo, 'a.*', 'i')` with flag support (eg. `'i'` for case insensitive match)
- H2: Same as Oracle
- SQLite: Same as MySQL

Even though I normally like how PostgreSQL does things, in this case I think it's a bit weird and prefer the Oracle syntax (implement it as a function, with arguments, which also allows regexp flags).

Alternatively, if we use the MySQL (or PostgreSQL) syntax, we could encode the flags within the expression itself, just like Java does (eg. `foo(?im)` is a case-insensitive multiline match). Let's figure out which one we want, because once it's in the grammar we'll be stuck with it for a long time.

jfim commented on pinot-controller/src/main/resources/webapp/js/init.js:
```diff
@@ -98,6 +98,7 @@ var HELPERS = {
 
   executeQuery: function(query, traceEnabled, callback) {
     var url = "/pql?pql=" + query +"&trace=" + traceEnabled;
+    url = encodeURI(url);
```
Does that actually work? I would think you'd need to encode `query` instead

kishoreg replied:
It seemed to work but I did not test the traceEnabled part. But good point, I will encode the query part.

jfim replied:
Oh, interesting, `encodeURI` doesn't actually encode special characters like & and ?. In your case, it worked fine, but it means that if there is a query that contains a question mark, the question mark won't get encoded properly.

What you want in this case is `encodeURIComponent(query)` which will properly escape the query contents:

```
encodeURI('foo=bar?baz%')
"foo=bar?baz%25"
encodeURIComponent('foo=bar?baz%')
"foo%3Dbar%3Fbaz%25"
```

jfim commented on pinot-common/src/main/java/com/linkedin/pinot/pql/parsers/pql2/ast/RegexPredicateAstNode.java:
```diff
@@ -0,0 +1,83 @@
+        if (matchType == MatchType.LIKE) {
+          expr = expr.replace(".", "\\.");
+          expr = expr.replace("_", ".");
+          expr = expr.replaceAll("%", ".*?");
```
You also need to escape regular expression atoms, eg. `foo like 'abc (%)'` won't expand correctly.

kishoreg replied:
you mean escape ( [ ^ etc?.  I am wondering if we should just quote all segments before and after characters % and _?

jfim replied:
Yeah, escape the ( [ and such atoms.

jfim commented:
For the integration test, you can use `REGEXP_LIKE` in H2 (http://www.h2database.com/html/functions.html#regexp_like) as part of the query comparison/generation. I can help you with that part.

kishoreg commented:
I had trouble getting the same results as H2 in our integration test. REGEXP_LIKE in H2 uses pattern.compile.find() method instead of pattern.compile.matches(). I feel matches is the right thing to do but for now I have used find() to match with H2. Its not clear whats the behavior in other DB's.

Jackie-Jiang commented on pinot-common/src/main/java/com/linkedin/pinot/pql/parsers/pql2/ast/RegexPredicateAstNode.java:
```diff
@@ -0,0 +1,73 @@
+      throw new Pql2CompilationException("IN predicate has no identifier");
+    }
+
+    TreeSet<String> values = new TreeSet<>();
```
Any specific reason why we use TreeSet instead of HashSet?

kishoreg replied:
not really. changing it to hashset

Jackie-Jiang commented on pinot-core/src/main/java/com/linkedin/pinot/core/common/predicate/RegexPredicate.java:
```diff
@@ -18,19 +18,26 @@
 
 public class RegexPredicate extends Predicate {
-
+  String regex;
```
We have decided to use the convention _regex for member variable.

Jackie-Jiang commented on pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/RegexPredicateEvaluator.java:
```diff
@@ -0,0 +1,82 @@
+    for (int i = 0; i < length; i++) {
+      int dictId = dictionaryIds[i];
+      String value = dictionary.getStringValue(dictId);
+      if (pattern.matcher(value).matches()) {
```
matches() or find()?

kishoreg replied:
it was changed to find. Not sure why the PR is not reflecting it. will update the PR again.

Jackie-Jiang commented on pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/OfflineClusterIntegrationTest.java:
```diff
@@ -51,6 +51,7 @@
 /**
  * Integration test that converts Avro data for 12 segments and runs queries against it.
  */
+@Test
```
We don't need this annotation to mark it as a test class

kishoreg replied:
for some reason, eclipse needs that.

Jackie-Jiang commented on pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/QueryGenerator.java:
```diff
@@ -799,7 +799,44 @@ public QueryFragment generatePredicate(String columnName) {
+
+        String pql = String.format(" REGEXP_LIKE(%s, '%s')", columnName, pqlRegexBuilder.toString());
+        String sql = String.format(" REGEXP_LIKE(%s, '%s', 'i')", columnName, sqlRegexBuilder.toString());
+        System.out.println("Value:" + value + " pql:" + pql + " sql:" + sql);
```
Do not do sout because it might polute the test log and break Travis-CI for the number of output lines limit.

mayankshriv commented on pinot-common/src/main/java/com/linkedin/pinot/pql/parsers/pql2/ast/RegexPredicateAstNode.java:
```diff
@@ -0,0 +1,72 @@
+public class RegexPredicateAstNode extends PredicateAstNode {
+  private String _identifier;
+
+  public RegexPredicateAstNode() {
```
Is this needed?

kishoreg replied:
removed

mayankshriv commented on pinot-common/src/main/java/com/linkedin/pinot/pql/parsers/pql2/ast/RegexPredicateAstNode.java:
```diff
@@ -0,0 +1,72 @@
+
+    String[] valueArray = values.toArray(new String[values.size()]);
+    FilterOperator filterOperator = FilterOperator.REGEX;
+    List<String> value = Collections.singletonList(StringUtil.join("\t\t", valueArray));
```
static final for "\t\t"

mayankshriv commented on pinot-common/src/main/java/com/linkedin/pinot/pql/parsers/utils/PQLParserUtils.java:
```diff
@@ -124,4 +124,25 @@ public static void decorateWithMapReduce(JSONObject jsonObj,
+   * @param expr
+   * @return
+   */
+  public static String convertLikeExpressionToJavaRegex(String expr) {
```
Does this have unit test?

kishoreg replied:
I will delete this method as I removed the support for LIKE syntax

mayankshriv commented on pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/RegexPredicateEvaluator.java:
```diff
@@ -0,0 +1,82 @@
+  private Dictionary dictionary;
+  private Pattern pattern;
+
+  public RegexPredicateEvaluator(RegexPredicate predicate, Dictionary dictionary) {
```
Just a heads up, there's another PR in-flight which requires no-dictionary support for predicate evaluators.

kishoreg replied:
Good point. Should be easy for regex since we don't have to differentiate based on datatype

kishoreg commented:
I realized that we had REGEX enum already defined in request thrift file. Since no one has used this yet, I thought this was the right time to fix that as well. I changed REGEX to REGEXP_LIKE and renamed all predicate classed to reflect the same.
